api/models/trigger.go

```diff
@@ -0,0 +1,175 @@
+package models
+
+import (
+	"errors"
+	"net/http"
+	"time"
+
+	"github.com/fnproject/fn/api/id"
+	"github.com/go-openapi/strfmt"
+)
+
+//go:generate jsonenums -type=TriggerType
+type TriggerType int
+
+const (
+	Unknown TriggerType = iota
+	HTTP
+)
+
+type Trigger struct {
+	ID          string          `json:"id" db:"id"`
```

Comment:
Have we decided to ditch routers.Headers  ? 

Also should "type" be validated (e.g. dynamically based on pluggable trigger types)

Reply:
I guess my thoughts are : 

for HTTP there are a clear set of possible  configuration values: e.g.: 

* CORS configuration 
* HTTP method selectors (in future) (e.g. different Fns for PUT/GET) 
* Current "Jack in my headers" behaviour (which may be a bit broken in many ways) 

Was wondering about the relative mertits of : 

* Hard coding these on the Trigger Object as sub-classes (e.g. HttpTrigger < Trigger) via Swagger polymorphism 
* Setting up some independent Annotation specs for these 

The former is obviously tighter but scaling this up to other classes of event source might be trigger (CronTrigger looks easy, but QueueTrigger looks harder) - the latter feels too loose in general - especially for HTTP having unbound/typed properties for core values seems wrong.

Reply:
> Have we decided to ditch routes.Headers ?

yes. we can add it back if needed but it's not a required feature

> Also should "type" be validated (e.g. dynamically based on pluggable trigger types)

yea i think it should just be a flexible string, then we allow custom triggers out of the box. extensions could possibly validate a limited set (say, a service that only supports x and y), but we probably shouldn't in fn core

Reply:
I don't like the idea of a flexible string for the trigger type. (Http, HTTP, http, etc, etc, needless pain).

I'm not sure how there could be a 'custom' trigger without it existing in a code base, somehow. In the json transfer the names are used ('HTTP' for example), it's only internally the int value is used. So there's not much danger of someone porting something over and it changing type.

I think annotation specs should come in as soon as we've nailed down what they need to be for HTTP. They should be open for extension though.